### PR TITLE
RDM-7283: Introduce Validation Literals more Specific than just '[[AN…

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDM-7283

### Change description ###

Currently, we are supporting [[ANY]] literal to allow for any value of any type in a field in the actual response details (headers, body fields, etc). But this is a feature that mandates presence of a value in the field, no matter the type.

Replace [[ANY]] with relevant  literals to allow nulls and not nulls respectively.

**Does this PR introduce a breaking change?** (check one with "x")
Need to replace all [[ANY]] keywords with relevant keywords in feature files in all the projects

```
[Y ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    ANY_NOT_NULLABLE,
+    ANY_STRING_NOT_NULLABLE,
+    ANY_NUMBER_NOT_NULLABLE,
+    ANY_INTEGER_NOT_NULLABLE,
+    ANY_FLOATING_NOT_NULLABLE,
+    ANY_DATE_NOT_NULLABLE,
+    ANY_TIMESTAMP_NOT_NULLABLE,
+    ANY_OBJECT_NOT_NULLABLE,
+
+    NOT_NULL;
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java]: Would this be among the options? We already have ANY_NOT_NULLABLE.

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 2]
```diff
@@ -226,6 +245,39 @@ private Boolean canAcceptNullFor(Object expectedValue) {
...
+        return ANY_INTEGER_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_DATE_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_FLOATING_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_NUMBER_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_STRING_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_OBJECT_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_TIMESTAMP_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANY_NULLABLE.getValue().equalsIgnoreCase(strExpectedValue) ||
+                ANYTHING_IF_EXISTS.getValue().equalsIgnoreCase(strExpectedValue);
+    }
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: This and notNullable(..) method could be implemented more efficiently. We can have a boolean nullable field in the enum elements, pass a value in element constructors, and then here, we can get the element from strExpectedValue with a method like MapPlaceholder.getByValue(String value) from the enum type and check with its isNullable() method.

File: src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package uk.gov.hmcts.befta.util;
+
+public enum  MapPlaceholders {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java]: The name is a bit obscure.. How about, ExpectedValuePlaceholder? (Singular, not plural.)

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 4]
```diff
@@ -7,10 +7,27 @@
...
+import static uk.gov.hmcts.befta.util.MapPlaceholders.ANY_OBJECT_NULLABLE;
+import static uk.gov.hmcts.befta.util.MapPlaceholders.ANY_STRING_NOT_NULLABLE;
+import static uk.gov.hmcts.befta.util.MapPlaceholders.ANY_STRING_NULLABLE;
+import static uk.gov.hmcts.befta.util.MapPlaceholders.ANY_TIMESTAMP_NOT_NULLABLE;
+import static uk.gov.hmcts.befta.util.MapPlaceholders.ANY_TIMESTAMP_NULLABLE;
+import static uk.gov.hmcts.befta.util.MapPlaceholders.NOT_NULL;
 
-    private static final String ANY = "[[ANY]]";
-    private static final String NOT_NULL = "[[NOT_NULL]]";
+public class MapVerifier {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: In this class, we need to also check the types of the actual values against the types mandated by expected value place holders. For example, if the expected value is [[ANY_TIMESTAMP_NOT_NULLABLE]], we have to make sure there is a value in the actual field, and it is indeed in timestamp format.

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java (hunk 2 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: Changed the implementation

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java (hunk 4 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: Changed implementation

File: src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java (hunk 3 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java]: Changed the name

File: src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java (hunk 1 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapPlaceholders.java]: Removed NOT_NULL

File: src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    ExpectedValuePlaceholder(boolean nullable) {
+        this.nullable = nullable;
+    }
+
+    public boolean isNullable() {
+        return this.nullable;
+    }
+
+    public boolean isValidFormat(Object actualObject) {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java]: could we rename this so accepts(Object actualValue)
to read a code like if(somePlaceholder.accepts(actualValue))

File: src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        if (actualObject != null) {
+            try {
+                switch (this) {
+                    case ANY_STRING_NULLABLE:
+                    case ANY_STRING_NOT_NULLABLE:
+                        valid &= actualObject instanceof String;
+                        break;
+                    case ANY_INTEGER_NOT_NULLABLE:
+                    case ANY_INTEGER_NULLABLE:
+                        Integer.parseInt(actualObject.toString());
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java]: This will throw a numberformatexception without actually returning a false value.

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 7]
```diff
@@ -226,6 +231,19 @@ private Boolean canAcceptNullFor(Object expectedValue) {
...
-        return !NOT_NULL.equalsIgnoreCase((String) expectedValue);
+        return nullable((String) expectedValue);
+    }
+
+    private boolean nullable(String strExpectedValue) {
+        ExpectedValuePlaceholder expectedValuePlaceholder = ExpectedValuePlaceholder.getByValue(strExpectedValue);
+        return expectedValuePlaceholder != null ?  expectedValuePlaceholder.isNullable(): true;
+    }
+
+    private boolean isValidFormat(String strExpectedValue, Object actualValue) {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: This doesn't sound like logically correct.

File: src/test/java/uk/gov/hmcts/befta/util/MapVerifierTest.java [hunk 8]
```diff
@@ -126,6 +143,189 @@ public void shouldVerifySimpleMapOfAcceptableContentWithWildcards() {
...
+        actualBody.put("callbackErrors", null);
+        actualBody.put("callbackWarnings", "Test Warnings");
+
+        MapVerificationResult result = new MapVerifier("actualResponse", 0).verifyMap(expected, actual);
+        Assert.assertEquals(0, result.getAllIssues().size());
+        Assert.assertTrue(result.isVerified());
+    }
+
+    @Test
+    public void shouldVerifyContentWithWildcardAnythingNullable() {
```
Comment (MSancaktutar) [inline src/test/java/uk/gov/hmcts/befta/util/MapVerifierTest.java]: Could it be good to split these tests into more granular ones each testing a particular placeholder for full conditional coverage?

File: src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java (hunk 5 above)
Comment (MSancaktutar) [reply src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java]: Otherwise nice one.

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 9]
```diff
@@ -226,6 +231,19 @@ private Boolean canAcceptNullFor(Object expectedValue) {
         if (!(expectedValue instanceof String)) {
             return Boolean.FALSE;
         }
-        return !NOT_NULL.equalsIgnoreCase((String) expectedValue);
+        return nullable((String) expectedValue);
+    }
+
+    private boolean nullable(String strExpectedValue) {
+        ExpectedValuePlaceholder expectedValuePlaceholder = ExpectedValuePlaceholder.getByValue(strExpectedValue);
+        return expectedValuePlaceholder != null ?  expectedValuePlaceholder.isNullable(): true;
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: pls avoid use of negation in logic checks.

Comment (MSancaktutar) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: return expectedValuePlaceholder == null ?  true: expectedValuePlaceholder.isNullable();

This reads more smoothly with same effect.

Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: Ok

File: src/test/java/uk/gov/hmcts/befta/util/MapVerifierTest.java (hunk 8 above)
Comment (kiran-yenigala-hmcts) [reply src/test/java/uk/gov/hmcts/befta/util/MapVerifierTest.java]: OK.. will do that change

File: src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java (hunk 5 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/ExpectedValuePlaceholder.java]: Changed to accepts(Object actualValue)

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java (hunk 7 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: I didn't understand.. what is the problem in this?

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 10]
```diff
@@ -122,15 +121,15 @@ private MapVerificationResult verifyMap(String fieldPrefix, Map<String, Object>
                         //
                     } else if (expectedValue == null) {
                         badValueMessages.add("Must be null: " + commonKey);
-                    } else if (actualValue == null) {
+                    } else if (actualValue == null && !nullable((String) expectedValue)) {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: else if (actualValue == null && !nullable((String) expectedValue))
--> else if (actualValue == null && isNoneNullablePlaceholder((String) expectedValue))

and 

private boolean isNoneNullablePlaceholder(String strExpectedValue) {
        ExpectedValuePlaceholder expectedValuePlaceholder = ExpectedValuePlaceholder.getByValue(strExpectedValue);
        return expectedValuePlaceholder != null &&  !expectedValuePlaceholder.isNullable();
}

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 11]
```diff
@@ -122,15 +121,15 @@ private MapVerificationResult verifyMap(String fieldPrefix, Map<String, Object>
...
                         badValueMessages.add("Must not be null: " + commonKey);
                     } else if (!(expectedValue instanceof Map && actualValue instanceof Map)) {
                         if (expectedValue instanceof Collection<?> && actualValue instanceof Collection<?>) {
                             collectBadValueMessagesFromCollection(fieldPrefix + "." + commonKey, commonKey,
                                     (Collection<?>) expectedValue,
                                     (Collection<?>) actualValue, currentDepth, maxMessageDepth, badValueMessages);
                         } else {
-                            Object outcome = compareValues(fieldPrefix, commonKey, expectedValue, actualValue,
+                            Object outcome = compareValues(commonKey, expectedValue, actualValue,
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: Why this change? A fix by opportunity?

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java [hunk 12]
```diff
@@ -186,12 +184,16 @@ private void collectBadValueMessagesFromCollection(String fieldPrefix, String fi
...
+    private Object compareValues(String commonKey, Object expectedValue,
             Object actualValue, int currentDepth, int maxMessageDepth) {
         boolean justCompare = currentDepth > maxMessageDepth;
-        if (expectedValue instanceof String && ANY.equalsIgnoreCase((String) expectedValue)) {
+        if (expectedValue instanceof String &&
+                ANYTHING_IF_EXISTS.getValue().equalsIgnoreCase((String) expectedValue)) {
             return Boolean.TRUE;
-        } else if (expectedValue == actualValue) {
+        } else if(expectedValue instanceof String && isValidFormat((String) expectedValue, actualValue)) {
+            return Boolean.TRUE;
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: these 2 checks could be combined in one method like
private boolean actualAcceptedByPlaceholder(String strExpectedValue, Object actualValue) {

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java (hunk 10 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: changed method name

File: src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java (hunk 12 above)
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/befta/util/MapVerifier.java]: Changed

**Comment by MSancaktutar** [comment]
Merged myself after all new unit tests passing against this change and 2 consecutive builds going green on data store.
